concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -114,11 +115,74 @@ placeholderSectionHeader = SectionHeader 0 0 P1 (BlockHash minBound) 0 0 0 0
--- location is at the end of the the written section.
-writeSection ::
-    (MonadIO m) =>
+-- |Open a file handle for writing a chunk.
+initialHandle :: FilePath -> IO Handle
+initialHandle p = do
+    createDirectoryIfMissing True (takeDirectory p)
+    openBinaryFile p WriteMode
+
+-- |Export a database in V3 format, as a collection of block file chunks, given the data directory
+-- root and the export directory root.
+exportDatabaseV3 ::
+    -- |Data directory
+    FilePath ->
+    -- |Export directory
+    FilePath ->
+    -- |Chunk size
+    Word64 ->
+    IO ()
+exportDatabaseV3 dbDir outDir chunkSize = do
+  createDirectoryIfMissing True outDir
+  bracket (openFile (outDir </> "blocks.idx") AppendMode) hClose
+    (\indexHdl -> exportSections dbDir outDir indexHdl chunkSize 0)
+
+-- |Export a database section for each genesis index.
+exportSections ::
+    -- |Database directory
+    FilePath ->
+    -- |Export directory
+    FilePath ->
+    -- |Index file handle
+    Handle ->
+    -- |Chunk size
+    Word64 ->
+    -- |Next genesis index to export
+    GenesisIndex ->
+    IO ()
+exportSections dbDir outDir indexHdl chunkSize genIndex = do
+    let treeStateDir = dbDir </> "treestate-" ++ show genIndex
+    -- Check if the database exists for this genesis index
+    dbEx <- doesPathExist $ treeStateDir </> "data.mdb"
+    if dbEx
+        then
+            openReadOnlyDatabase treeStateDir >>= \case
+                Nothing -> putStrLn "Tree state database could not be opened."
+                Just (VersionDatabaseHandlers (dbh :: DatabaseHandlers pv ())) -> do
+                    evalStateT
+                        ( do
+                            mgenFinRec <- resizeOnResized $ readFinalizationRecord 0
+                            forM_ mgenFinRec $ \genFinRec -> do
+                                let genHash = finalizationBlockPointer genFinRec
+                                    startHeight = 1
+                                writeChunks
+                                    genIndex
+                                    (demoteProtocolVersion (protocolVersion @pv))
+                                    genHash
+                                    startHeight
+                                    outDir
+                                    indexHdl
+                                    chunkSize
+                        )
+                        (DBState dbh)
+                    closeDatabase dbh
+                    exportSections dbDir outDir indexHdl chunkSize (genIndex + 1)
+        else putStrLn ("The tree state database does not exist at " ++ treeStateDir)
+
+-- |Write a database section as a collection of chunks in the specified directory. The exported
+-- chunks will be accompanied by an index file mapping chunk filenames to the block ranges stored in
+-- them. Each line in index file has format FILENAME,FIRSTBLOCK,LASTBLOCK\n
```

Comment:
There is missing documentation as to what is in each chunk. That is, how the finalization records in the chunk compare to the blocks in it.